### PR TITLE
azure features - api, website, function (heina verify diff, this is not the real PR) real PR has target branch of azure

### PR DESCRIPTION
**Pull request by heina-chang**
### Requirements

* Filling out the template is required. Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.
* All new code requires tests to ensure against regressions

### CODE REVIEW

File: builds/delete-serverless-service-build-pack/Jenkinsfile [hunk 1]
```diff
@@ -1269,3 +1325,45 @@ def parseJsonMap(jsonString) {
...
+  ]
+
+  azureUtil.setAzureVar(serviceInfo)
+	withCredentials([
+			[$class: 'UsernamePasswordMultiBinding', credentialsId: 'AZ_PASSWORD', passwordVariable: 'AZURE_CLIENT_SECRET', usernameVariable: 'UNAME'],
+			[$class: 'UsernamePasswordMultiBinding', credentialsId: 'AZ_CLIENTID', passwordVariable: 'AZURE_CLIENT_ID', usernameVariable: 'UNAME'],
+			[$class: 'UsernamePasswordMultiBinding', credentialsId: 'AZ_TENANTID', passwordVariable: 'AZURE_TENANT_ID', usernameVariable: 'UNAME'],
+			[$class: 'UsernamePasswordMultiBinding', credentialsId: 'AZ_SUBSCRIPTIONID', passwordVariable: 'AZURE_SUBSCRIPTION_ID', usernameVariable: 'UNAME']
+	]) {
+    def repo_name = "jazz_azure-create-service"
```
Comment (devsatishm) [inline builds/delete-serverless-service-build-pack/Jenkinsfile]: What is this used for (during deletion)?

Comment (heina-chang) [reply builds/delete-serverless-service-build-pack/Jenkinsfile]: this is to delete azure resources via that nodejs SDK project.  So we retrieve azure principle from jenkins credentials and call our nodejs project to delete resesources

File: builds/delete-serverless-service-build-pack/Jenkinsfile [hunk 2]
```diff
@@ -1269,3 +1325,45 @@ def parseJsonMap(jsonString) {
 	lazyMap.putAll(jsonObj)
 	return lazyMap
 }
+
+def isAzure() {
+  return  service_config['platform'] != null && service_config['platform'] == "azure"
+}
+
+def deleteResourceByTag(env) {
```
Comment (devsatishm) [inline builds/delete-serverless-service-build-pack/Jenkinsfile]: Function contains business logic related to azure (like populating stackName). Can all of this move to azure module (groovy util)?

File: builds/jazz-build-module/azure-function-deploy-module.groovy [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+  this.configLoader = configLoader
+  this.utilModule = utilModule
+  this.scmModule = scmModule
+  this.events = events
+  this.azureUtil = azureUtil
+
+}
+
+
+def createFunction(serviceInfo){
```
Comment (devsatishm) [inline builds/jazz-build-module/azure-function-deploy-module.groovy]: Need function documentation (purpose, input etc.)

File: builds/jazz-build-module/azure-function-deploy-module.groovy [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+
+          def output = azureUtil.invokeAzureService(data, "getMasterKey")
+          masterKey = output.data.result.key
+          deployFunction(data, zip, type)
+          sendAssetCompletedEvent(serviceInfo, assetList)
+          return masterKey
+        } catch (ex) {
+          echo "error occur $ex, rollback starting..."
+          deleteResourceByTag(serviceInfo)
+          error "Failed creating azure function $ex"
```
Comment (devsatishm) [inline builds/jazz-build-module/azure-function-deploy-module.groovy]: Every error (jenkins error) should be caught and FAILURE events must be sent before the jenkins workflow errors out. Unless this happens, we can't mark the deployment as failed in jazz metadata.

File: builds/delete-serverless-service-build-pack/Jenkinsfile (hunk 2 above)
Comment (heina-chang) [reply builds/delete-serverless-service-build-pack/Jenkinsfile]: fixed https://github.com/dev9com/jazz/pull/81/files, will merge once pass travis green

File: builds/jazz-build-module/azure-function-deploy-module.groovy (hunk 3 above)
Comment (heina-chang) [reply builds/jazz-build-module/azure-function-deploy-module.groovy]: fixed https://github.com/dev9com/jazz/pull/81/files

File: builds/jazz-build-module/azure-function-deploy-module.groovy [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+
+    events.sendCompletedEvent('CREATE_ASSET', null, utilModule.generateAssetMap(serviceInfo.serviceCatalog['platform'], id, assetType, serviceInfo.serviceCatalog), serviceInfo.envId)
+
+  }
+
+  if (assetList) {
+    for (item in assetList) {
+      def id = item.azureResourceId
+      def assetType = item.type
+      events.sendCompletedEvent('CREATE_ASSET', null, utilModule.generateAssetMap(serviceInfo.serviceCatalog['platform'], id, assetType, serviceInfo.serviceCatalog), serviceInfo.envId)
```
Comment (devsatishm) [inline builds/jazz-build-module/azure-function-deploy-module.groovy]: We are sending CREATE_ASSET event twice in this function. Is this expected?

Comment (heina-chang) [reply builds/jazz-build-module/azure-function-deploy-module.groovy]: yes.  
The first loop return us the assets created by this service.  The end loop is assets created by this service but not an azure resource such as database, table, queue, blob.
The loop assetList is track by this jenkins file but jenkins file does not know an azure resource is newly created or an existing which is to cover the requirement of allowing user enter existing resources.

**Comment by heina-chang** [comment]
dups with #943

**Pull request closed without merging**